job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java

```diff
@@ -0,0 +1,606 @@
+
+        finishedTimeSlices = Optional.empty();
+        jobFinished = Optional.empty();
+    }
+
+    /**
+     * Test hook to allow control of when ticks are emitted.
+     */
+    void setTickScheduler(rx.Scheduler scheduler) {
+        tickScheduler = scheduler;
+    }
+
+    /**
+     * Test hook to broadcast when the job scheduler has finished all work for a time slice. This includes not only
+     * executing jobs that are scheduled for a particular time slice but also post-job execution clean up.
+     */
+    void setTimeSlicesSubject(PublishSubject<Date> timeSlicesSubject) {
+        finishedTimeSlices = Optional.of(timeSlicesSubject);
+    }
+
+    /**
+     * Test hook to broadcast when jobs finish executing.
+     */
+    void setJobFinishedSubject(PublishSubject<JobDetails> subject) {
+        jobFinished = Optional.of(subject);
+    }
+
+    @Override
+    public void registerJobFactory(String jobType, Func1<JobDetails, Completable> factory) {
+        jobFactories.put(jobType, factory);
+    }
+
+    @Override
+    public Single<JobDetails> scheduleJob(String type, String name, Map<String, String> parameter, Trigger trigger) {
+        if (System.currentTimeMillis() >= trigger.getTriggerTime()) {
+            return Single.error(new RuntimeException("Trigger time has already passed"));
+        }
+        String lockName = "org.hawkular.metrics.scheduler.queue." + trigger.getTriggerTime();
+        return lockManager.acquireLock(lockName, "scheduling", 5)
+                .map(acquired -> {
+                    if (acquired) {
+                        UUID jobId = UUID.randomUUID();
+                        return new JobDetails(jobId, type, name,parameter, trigger);
+                    }
+                    throw new RuntimeException("Failed to acquire scheduling lock [" + lockName + "]");
+                })
+                .toSingle()
+                .flatMap(details -> Completable.merge(insertIntoJobsTable(details), updateScheduledJobsIndex(details))
+                            .andThen(Single.just(details))
+                );
+    }
+
+    @Override
+    public void start() {
+        running = true;
+        NavigableSet<Date> activeTimeSlices = new ConcurrentSkipListSet<>();
+        Set<UUID> activeJobs = new ConcurrentSkipListSet<>();
+
+        doOnTick(() -> {
+            logger.debug("Activating scheduler for [" + currentMinute().toDate() + "]");
+
+            Date timeSlice = currentMinute().toDate();
+            // TODO Figure out why updateActiveTimeSlices() with Completable doesn't work
+            // When using updateActiveTimeSlices() that returns a Completable, we get intermittent failures. Need
+            // understand why. Maybe it has something to do with converting from ListenableFuture to Observable to
+            // Completable.
+            updateActiveTimeSlices(currentMinute().toDate()).andThen(findTimeSlices())
```

Comment:
That's correct. The same information will be returned to each node, and that is expected. The concurrency issue you debugged actually is specific to single node as it involves the ```activeTimeSlices``` cache which is a local variable that keeps track of the active time slices for the running node. We want to allow multiple job scheduler nodes to run concurrently, executing jobs over the same set of time slices. Locks are the coordination mechanism here.